### DISCUSSION TITLE
WPF如何使用listbox进行MVVM导航。

### DISCUSSION THREAD

**Discussion by ICANfly0402** [BYJRK/DotNet-Discussions#320] [MVVM]
你好，我在B站关注了你，正在学习WPF。看了你的视频关于WPF使用MVVM进行导航，我想请教一下WPF如何使用listbox进行MVVM导航，我看了你罗列的几种方法都是基于button导航的。我尝试修改SingletonSean的demo，使用listbox能导航但是viewmodel之间无法传输userservice数据。请问是否有好的方法呢？谢谢

**MainWindow.xaml：**
 <Window x:Class="SingletonSean.MainWindow"
        xmlns="http://schemas.microsoft.com/winfx/2006/xaml/presentation"
        xmlns:x="http://schemas.microsoft.com/winfx/2006/xaml"
        xmlns:d="http://schemas.microsoft.com/expression/blend/2008"
        xmlns:mc="http://schemas.openxmlformats.org/markup-compatibility/2006"
        xmlns:local="clr-namespace:SingletonSean.Views"
        xmlns:vm="clr-namespace:SingletonSean.ViewModels"
        Title="SingletonSean's MVVM Nagivation"
        Width="800"
        Height="450"
        d:DataContext="{d:DesignInstance Type=vm:MainViewModel}"
        WindowStartupLocation="CenterScreen"
        mc:Ignorable="d">

    <DockPanel>
        <ListBox x:Name="PagesListBox"
                Width="200"
                Background="Gray"
                DisplayMemberPath="UserName"
                DockPanel.Dock="Left"
                ItemsSource="{Binding Pages}"
                SelectedItem="{Binding CurrentViewModel, Mode=TwoWay,UpdateSourceTrigger=PropertyChanged}" />
        <ContentControl Content="{Binding CurrentViewModel}">
            <ContentControl.Resources>
                <DataTemplate DataType="{x:Type vm:LoginViewModel}">
                    <local:Login />
                </DataTemplate>
                <DataTemplate DataType="{x:Type vm:HomeViewModel}">
                    <local:Home />
                </DataTemplate>
            </ContentControl.Resources>
        </ContentControl>
    </DockPanel>
</Window>

**MainViewModel.cs：**
namespace SingletonSean.ViewModels
{
    public partial class MainViewModel : ViewModelBase
    {
        private readonly NavigationService navigator;

        [ObservableProperty]
        private ObservableCollection<ViewModelBase> pages;

        [ObservableProperty]
        private ViewModelBase? currentViewModel;

        public MainViewModel(NavigationService navigationService, HomeViewModel hm, LoginViewModel log)
        {
            Pages = new ObservableCollection<ViewModelBase>() { log, hm };

            navigator = navigationService;

            navigator.CurrentViewModelChanged += () =>
            {
                CurrentViewModel = navigationService.CurrentViewModel;
            };

            // navigate to login page at first
            navigator.NavigateTo<LoginViewModel>();
        }
    }
}

**Comment by ICANfly0402** [comment]
我补充一下，我是说去切换listitem时无法传输数据
![image](https://github.com/user-attachments/assets/7ffeaa7e-d008-4314-97bf-791f3676c01c)
![image](https://github.com/user-attachments/assets/7a4cea06-5ad8-4e0d-9039-46522ad22044)

**Comment by KiyanYang** [comment]
【如何使用 CommunityToolkit.Mvvm 中的 Messenger 来进行 ViewModel 之间的通信-哔哩哔哩】 https://b23.tv/GOyayfL

**Comment by ICANfly0402** [comment]
十月的demo中使用全局USERSERVICE共享数据，使用listbox切换页面导航的话怎样也能用上USERSERVICE共享数据呢
